pkg/logs/internal/decoder/auto_multiline_detection/pattern_table.go

```diff
@@ -0,0 +1,146 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package automultilinedetection contains auto multiline detection and aggregation logic.
+package automultilinedetection
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/logs/internal/decoder/auto_multiline_detection/tokens"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type row struct {
+	tokens    []tokens.Token
+	label     Label
+	count     int64
+	lastIndex int64
+}
+
+// DiagnosticRow is a struct that represents a diagnostic view of a row in the PatternTable.
+type DiagnosticRow struct {
+	TokenString string
+	LabelString string
+	Count       int64
+	LastIndex   int64
+}
+
+// PatternTable is a table of patterns that occur over time from a log source.
+// This heuristic is used to prevent certain classes of false positives in multiline detection
+// as well as collect diagnostic data for fine tuning and debugging.
+// The patternt table is always sorted by the frequency of the patterns. When the table
+// becomes full, the least recently updated pattern is evicted.
+type PatternTable struct {
+	table                           []*row
+	index                           int64
+	maxTableSize                    int
+	matchThreshold                  float64
+	dontAggregateUnmatchedTopFormat bool
+}
+
+// NewPatternTable returns a new PatternTable heuristic.
+func NewPatternTable(maxTableSize int, matchThreshold float64, dontAggregateUnmatchedTopFormat bool) *PatternTable {
+	return &PatternTable{
+		table:                           make([]*row, 0, maxTableSize),
+		index:                           0,
+		maxTableSize:                    maxTableSize,
+		matchThreshold:                  matchThreshold,
+		dontAggregateUnmatchedTopFormat: dontAggregateUnmatchedTopFormat,
+	}
+}
+
+func (p *PatternTable) insert(tokens []tokens.Token, label Label) int {
+	p.index++
+	foundIdx := -1
+	for i, r := range p.table {
+		if isMatch(r.tokens, tokens, p.matchThreshold) {
+			r.count++
+			r.label = label
+			r.lastIndex = p.index
+			foundIdx = i
+			break
+		}
+	}
+
+	if foundIdx == 0 {
+		return foundIdx
+	}
+
+	if foundIdx > 0 {
+		return p.siftUp(foundIdx)
+	}
+
+	// If the table is full, make room for a new entry
+	if len(p.table) >= p.maxTableSize {
```

Comment:
Using `>=` means that we allow the table to grow pass its max table size. Should we use `==` instead?